File: src/cz/vutbr/fit/interlockSim/objects/cells/InOut.java
/* Brno University of Technology
 * Faculty of Information Technology
 * 
 * BSc Thesis	2006/2007
 * 
 * Railway Interlocking Simulator
 * 
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.objects.cells;

import java.util.EnumSet;
import java.util.Set;

import cz.vutbr.fit.interlockSim.objects.cells.RailSemaphore.Signal;
import cz.vutbr.fit.interlockSim.objects.paths.PathElement;
import cz.vutbr.fit.interlockSim.sim.PathSeparatorChangeException;

/**
 * predstavuje spojeni s externi zeleznicni siti
 *
 */
public class InOut extends OrientedNodeCell {
	private String name;
	private final RailSemaphore inSemaphore;
	private final RailSemaphore outSemaphore;
	
	/**
	 * @param name
	 * @param orientation
	 * @param spatialType
	 */
	public InOut(String name, Boolean orientation, SpatialType spatialType) {
		super(orientation, spatialType);
		this.name = name;
		this.inSemaphore = new RailSemaphore(!orientation, spatialType); 
		assert inSemaphore.direction() == Segment.anti(direction()); 
		this.outSemaphore = RailSemaphore.getConstantInstance(orientation, spatialType, Signal.FREE);
		setName(name);
	}
	
	public Set<Segment> joins() {
		return EnumSet.of(direction());
	}

	@Override
	public Segment getFollowingSegment(Segment from) {
		if (from == null) return direction();
		assert from == direction();
		return null;
	}
	
	/**
	 * @return semaphore on input
	 */
	public RailSemaphore getInSemaphore() {
		return inSemaphore;
	}
	
	/**
	 * @return name of place
	 */
	public String getName() {
		return name;
	}

	/**
	 * @return semaphore on output
	 */
	public RailSemaphore getOutSemaphore() {
		return outSemaphore;
	}
		
	private RailSemaphore getSemaphoreFor(Segment from, Segment to) {
		if (from == null && to == direction()) return inSemaphore;
		else if (to == null && from == direction()) return outSemaphore;
		return null;
	}
	
	private RailSemaphore getSemaphoreForWithExeption(Segment from, Segment to) throws PathSeparatorChangeException {
		final RailSemaphore sem = getSemaphoreFor(from, to);
		if (sem == null) throw new PathSeparatorChangeException(this);
		return sem;
	}
	
	public void setUpPath(Segment from, Segment to, double allowedSpeed) throws PathSeparatorChangeException {
		final RailSemaphore sem = getSemaphoreForWithExeption(from, to);
		sem.setSignal(Signal.forSpeed(allowedSpeed));
	}
	
	public void cancelPathSetup(Segment from, Segment to) throws PathSeparatorChangeException {
		final RailSemaphore sem = getSemaphoreForWithExeption(from, to);
		sem.setSignal(Signal.STOP);
	}
	
	public double allowedSpeed() {
		return PathElement.ABSOLUTE_MAX_SPEED;
	}
}

File: src/main/kotlin/cz/vutbr/fit/interlockSim/objects/cells/InOut.kt
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.objects.cells

import cz.vutbr.fit.interlockSim.exceptions.requireSimulation
import cz.vutbr.fit.interlockSim.objects.core.Cell
import cz.vutbr.fit.interlockSim.objects.core.anti

/**
 * predstavuje spojeni s externi zeleznicni siti
 *
 */
class InOut(
	name: String,
	orientation: Boolean,
	spatialType: Cell.SpatialType
) : OrientedNodeCell(orientation, spatialType) {
	private var name: String
	private val inSemaphore: RailSemaphore
	private val outSemaphore: RailSemaphore

	init {
		this.name = name
		this.inSemaphore = RailSemaphore(!orientation, spatialType)
		requireSimulation(inSemaphore.direction() == anti(direction())) {
			"In semaphore direction must be anti-parallel to InOut direction"
		}
		this.outSemaphore = RailSemaphore(orientation, spatialType)
		setName(name)
	}

	override fun joins(): Set<Cell.Segment> = setOf(direction())

	override fun getFollowingSegment(from: Cell.Segment?): Cell.Segment? {
		if (from == null) return direction()
		requireSimulation(from === direction()) { "Invalid segment: $from, expected: ${direction()}" }
		return null
	}

	/**
	 * @return semaphore on input
	 */
	fun getInSemaphore(): RailSemaphore = inSemaphore

	/**
	 * @return name of place
	 */
	override fun getName(): String = name

	/**
	 * Set the name of this InOut.
	 * Overrides parent to update the local name field (which shadows parent's name).
	 *
	 * @param name The new name for this InOut
	 */
	override fun setName(name: String) {
		this.name = name
		super.setName(name) // Also update parent's name for consistency
	}

	/**
	 * @return semaphore on output
	 */
	fun getOutSemaphore(): RailSemaphore = outSemaphore

	/**
	 * Implementation of asRailSemaphore() for InOut.
	 * Returns the output semaphore since InOut acts as an exit point.
	 */
	override fun asRailSemaphore(): RailSemaphore = outSemaphore

	/**
	 * Get the track connection direction for InOut.
	 *
	 * InOut connects to the railway network at `direction()` regardless of whether
	 * trains are entering or exiting. The track connection is bidirectional.
	 *
	 * ## Bidirectional Semantics
	 *
	 * InOut has two semaphores:
	 * - **inSemaphore**: faces `anti(direction())` - controls trains entering FROM external network
	 * - **outSemaphore**: faces `direction()` - controls trains exiting TO external network
	 *
	 * The track connection is ALWAYS at `direction()`:
	 * - **Entry path**: FROM null (outside at anti-direction) TO direction() (track) → uses inSemaphore
	 * - **Exit path**: FROM direction() (track) TO null (outside at anti-direction) → uses outSemaphore
	 *
	 * See [DynamicInOut.getSemaphoreFor] for the logic that determines which semaphore
	 * controls a path based on from/to segments.
	 *
	 * @return Track connection direction (same as output direction)
	 * @see DynamicInOut.getSemaphoreFor
	 * @see getInSemaphore
	 * @see getOutSemaphore
	 */
	fun getTrackConnectionDirection(): Cell.Segment = direction()
}
